squint.c

```diff
@@ -0,0 +1,1611 @@
+   for (scan = funcBegin; scan <= funcEnd; ++scan)
+      if (*scan == NOP13 && !is_const(scan)) return;
+
+   /* record frame variable in this context */
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if ((*scan & 0x0f2f0000) == 0x050b0000) { // (ldr|str)[b], [fp, #X]
+         int off = (*scan & 0xfff);
+         if  ((*scan & (1<<23)) == 0) off = -off;
+         for (i = 0; i < numReg; ++i) {
+            if (offset[i] == off) break;
+         }
+         if (i == numReg) {
+            offset[numReg++] = off;
+            if (numReg == MAX_REN_REG) break;
+         }
+      }
+   }
+
+   /* discard any frame variables that are not trivial */
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if ((*scan & 0xffffff00) == 0xe28b0000 || // add r0, fp, #X
+          (*scan & 0xffffff00) == 0xe24b0000) { // sub r0, fp, #X
+         int off = *scan & 0xff;
+         if ((*scan & 0xffffff00) == 0xe24b0000) {
+            off = -off;
+         }
+         for (i = 0; i < numReg; ++i) {
+            if (offset[i] == off) break;
+         }
+         if (i != numReg) {
+            --numReg;
+            for(; i < numReg; ++i)
+               offset[i] = offset[i+1];
+         }
+      }
+   }
+
+   if (numReg == 0) return;
+
+   /* create registers for frame vars */
+   for (scan = funcBegin; scan <= funcEnd; ++scan) {
+      scan = skip_nop(scan, 1);
+
+      if ((*scan & 0x0f2f0000) == 0x050b0000) { // (ldr|str)[b], [fp, #X]
+         int off = (*scan & 0xfff);
+         if  ((*scan & (1<<23)) == 0) off = -off;
+         for (i = 0; i < numReg; ++i) {
+            if (offset[i] == off) break;
+         }
+         if (i == numReg) continue; // this frame var not mapped
+         if ((*scan & 0x0f3f0000) == 0x051b0000) { // ldr[b] rX, [fp, #X]
+            *scan = 0xe1a00000 | (*scan & 0x0000f000) | (BR+i);
+         }
+         else { // str[b] rd, [fp, #X] -> mov rX, rd
+            *scan = 0xe1a00000 | ((BR+i) << 12) | ((*scan >> 12) & 0xf);
+         }
+      }
+   }
+
+   /* load frame vars into registers at top of function */
+   for (scan = funcBegin; *scan != NOP; ++scan);
+   for (i = 0; i < numReg; ++i) {
+      *scan++ = 0xe51b0000 | ((BR + i) << 12) |
+                ((offset[i] < 0) ? -offset[i] : (offset[i] | (1<<23)));
+   }
+}
+
+/**********************************************************/
+/********* Peephole optimization driver function **********/
+/**********************************************************/
+
+int squint_opt(int *begin, int *end)
+{
+   int optApplied = 0 ;
+   int *scan = begin;
+   int *tmpbuf = (int *) malloc((end-begin)*sizeof(int));
+
+   create_const_map(begin, end);
+   const_imm_opt(begin, end);
+
+   while (scan < end) {
+      if (*scan == 0xe92d4800 && !is_const(scan)) { // push {fp, lr}
+         int *funcBegin = scan;
+         int *funcEnd;
+         int numReturn = 0;
```

Comment:
`numReturn` seems to be unused (only used by Line 1507) but not read by other.

Reply:
This was needed for early development/debugging and I will remove it now.